### PR TITLE
Adds Node List For Multisite

### PR DESCRIPTION
**Pull request by ljmcgann**
Adds first commands for multisite openstack
support

### CODE REVIEW

File: esiclient/v1/node_baremetal.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+import logging
+
+import openstack
+from osc_lib.command import command
+from osc_lib import exceptions
+from osc_lib.i18n import _
+
+# Initialize and turn on debug logging
+openstack.enable_logging(debug=False)
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:21]: this almost definitely should not be here. I think you may have turned on logging for the entire demo server by the way, as I see logs if I just log in and run 'openstack project list' - could you turn that off?

File: setup.cfg [hunk 2]
```diff
@@ -35,6 +35,7 @@ openstack.esiclient.v1 =
     esi_node_network_detach = esiclient.v1.node_network:Detach
     esi_node_network_list = esiclient.v1.node_network:List
     esi_node_volume_attach = esiclient.v1.node_volume:Attach
+    esi_mdc_baremetal_node_list = esiclient.v1.node_baremetal:MDCBaremetalNodeList
```
Comment (tzumainn) [inline setup.cfg:38]: These are generally in alphabetical order, so esi_mdc would go above all the esi_node entries.

File: esiclient/v1/node_baremetal.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                   "--clouds flag."
+                   )
+        )
+
+        return parser
+
+    def take_action(self, parsed_args):
+
+        columns = ['UUID', 'Name', 'Instance UUID', 'Power State',
+                   'Provisioning State', 'Maintenance']
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:57]: We want the "data center" - essentially the cloud name used from clouds.yaml - as well

File: esiclient/v1/node_baremetal.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            '--clouds',
+            dest='clouds',
+            metavar='<clouds>',
+            nargs="+",
+            required=True,
+            help=_("Specify the cloud to use from clouds.yaml.")
+        )
+
+        parser.add_argument(
+            '--regions',
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:41]: it looks like region is something that's part of clouds.yaml - can it be taken from there?

File: esiclient/v1/node_baremetal.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+class MDCBaremetalNodeList(command.Lister):
+    """List baremetal nodes from multiple OpenStack instances"""
+
+    log = logging.getLogger(__name__ + ".List")
+
+    def get_parser(self, prog_name):
+        parser = super(MDCBaremetalNodeList, self).get_parser(prog_name)
+        parser.add_argument(
+            '--clouds',
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:32]: I think this is a reasonable option, but can we also pull from an environment variable? This way the user only has to set it once.

File: esiclient/v1/node_baremetal.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+        if len(parsed_args.clouds) != len(parsed_args.regions):
+            raise exceptions.CommandError(
+                "ERROR: Must specify one region per cloud.")
+
+        for i in range(len(parsed_args.clouds)):
+            cloud = openstack.connect(cloud=parsed_args.clouds[i],
+                                      region=parsed_args.regions[i])
+
+            nodes = cloud.list_machines()
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:68]: This method is new to me, but it looks like a pretty good standardization. Interesting!

**Review by tzumainn — CHANGES_REQUESTED** [review]
Good start! Don't forget unit tests :)

File: esiclient/v1/node_baremetal.py (hunk 4 above)
Comment (tzumainn) [reply esiclient/v1/node_baremetal.py:41]: I'd say that the user should have multiple clouds.yaml entries then, one for each region. Regardless if this is an option or not, we want the default to be as simple as possible for the end user, and having them list a region for each cloud they want to use isn't all that simple.

Comment (ljmcgann) [reply esiclient/v1/node_baremetal.py:41]: It seems to be part of a normal workflow to have one cloud entry with multiple regions listed, the documentation for using the openstack.connect library gives an example using such a setup

Comment (tzumainn) [reply esiclient/v1/node_baremetal.py:41]: In that case we should iterate over the regionName; if that's not possible then we should document that the CLI doesn't support multiple regions in a cloud. Having to list all the region names for multiple clouds each time a user runs the CLI isn't user-friendly, and that's the primary goal of this feature.

**Comment by ljmcgann** [comment]
I have changed the command such that if no --clouds and --regions parameters are passed, all nodes from all clouds,region pairs in clouds.yaml will be displayed.

File: esiclient/v1/node_baremetal.py [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        data = []
+
+
+        if parsed_args.clouds is None and parsed_args.regions is None:
+            cloud_regions = openstack.config.loader.OpenStackConfig().get_all_clouds()
+            
+            for c in cloud_regions:
+
+                #cloud = openstack.connect(config=c)
+                cloud = openstack.connect(cloud=c._name, region=c.config['region_name'])
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:62]: Is the name value retrievable from c.config? I'm slightly uncomfortable using an "_" attribute.

File: esiclient/v1/node_baremetal.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        parser.add_argument(
+            '--regions',
+            dest='regions',
+            metavar='<regions>',
+            nargs="+",
+            help=_("Specify the region to use from clouds.yaml. Must specify "
+                   "regions in order corresponding to clouds passed with "
+                   "--clouds flag."
+                   )
+        )
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:45]: Let's just get rid of the region flag; I don't see why anyone would ever use it, and it cleans up the code below

File: esiclient/v1/node_baremetal.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+
+    def take_action(self, parsed_args):
+
+        columns = ['Cloud', 'Region', 'UUID', 'Name', 'Instance UUID', 'Power State',
+                   'Provisioning State', 'Maintenance']
+        data = []
+
+
+        if parsed_args.clouds is None and parsed_args.regions is None:
+            cloud_regions = openstack.config.loader.OpenStackConfig().get_all_clouds()
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:57]: Assuming we get rid of the region parameter, can we change this to filter out the undesired clouds if parsed_args.clouds is passed in? That gets rid of the else statement, which currently leads to some logic duplication.

**Review by tzumainn — COMMENTED** [review]
This works much better, thanks! Comments inline; reply on the comment if you fix them. Also, there's pep8 errors and the unit tests will need to be updated.

File: esiclient/v1/node_baremetal.py (hunk 7 above)
Comment (ljmcgann) [reply esiclient/v1/node_baremetal.py:62]: it is not

File: esiclient/v1/node_baremetal.py (hunk 8 above)
Comment (ljmcgann) [reply esiclient/v1/node_baremetal.py]: done

File: esiclient/v1/node_baremetal.py (hunk 9 above)
Comment (ljmcgann) [reply esiclient/v1/node_baremetal.py]: done

**Comment by ljmcgann** [comment]
I removed the region flag and fixed the unit tests

File: esiclient/v1/node_baremetal.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+        data = []
+
+        cloud_filter = parsed_args.clouds or []
+
+        cloud_regions = openstack.config.loader.OpenStackConfig().\
+            get_all_clouds()
+
+        for c in cloud_regions:
+
+            if cloud_filter and c._name not in cloud_filter:
```
Comment (tzumainn) [inline esiclient/v1/node_baremetal.py:50]: cloud_filter has a default value, so no need for the first part of the and clause

File: esiclient/tests/unit/v1/test_node_baremetal.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+            ]
+        )
+        self.assertEqual(expected, results)
+        mock_config.assert_called_once()
+        assert mock_cloud.list_machines.call_count == 3
+
+    @mock.patch('esiclient.v1.node_baremetal.openstack.connect')
+    @mock.patch('esiclient.v1.node_baremetal.openstack.config.loader.'
+                'OpenStackConfig.get_all_clouds')
+    def test_take_action_list(self, mock_config, mock_connect):
```
Comment (tzumainn) [inline esiclient/tests/unit/v1/test_node_baremetal.py:100]: This test has the same name as the other one; they should have different names.

**Review by tzumainn — COMMENTED** [review]
A few more comments; also the pep8 check is still failing.

File: esiclient/v1/node_baremetal.py (hunk 10 above)
Comment (ljmcgann) [reply esiclient/v1/node_baremetal.py:50]: its default value is to be empty if you dont pass in the --clouds flag. This means that nothing will print

File: esiclient/tests/unit/v1/test_node_baremetal.py (hunk 11 above)
Comment (ljmcgann) [reply esiclient/tests/unit/v1/test_node_baremetal.py:100]: my bad, fixed

**Comment by ljmcgann** [comment]
addressed comments

**Comment by tzumainn** [comment]
I forgot one more rather big change; can you rename the new file mdc_node_baremetal.py and put it into a new directory - esiclient/v1/mdc/? It seems like it'll be useful to separate these out.

**Comment by ljmcgann** [comment]
moved the appropriate files to mdc directory

File: setup.cfg [hunk 12]
```diff
@@ -31,6 +31,7 @@ openstack.cli.extension =
     esiclient = esiclient.plugin
 
 openstack.esiclient.v1 =
+    esi_mdc_baremetal_node_list = esiclient.v1.mdc.mdc_node_baremetal.py:MDCBaremetalNodeList
```
Comment (tzumainn) [inline setup.cfg:34]: The ".py" means this doesn't work

**Review by tzumainn — COMMENTED** [review]
I see one more pep8 error:

./esiclient/tests/unit/v1/mdc/test_mdc_node_baremetal.py:27:9: H238  old style class declaration, use new style (inherit from `object`)
        class FakeConfig():

File: setup.cfg (hunk 12 above)
Comment (ljmcgann) [reply setup.cfg:34]: yup, fixed

**Review by tzumainn — APPROVED** [review]

**Comment by tzumainn** [comment]
Looks good - thanks!

**Pull request merged**